### DISCUSSION TITLE
Which labels can be repeated multiple times?

### DISCUSSION THREAD

**Discussion by noomorph** [allure-framework/allure2#2059] [Questions & Support] [answered]
There are many built-in labels, and it's obvious that some can be repeated only once (like `severity`) and some can be repeated as many times as you want (like `tag`).

So far, I can think about this list: `allureId`, `package`, `testClass`, `testMethod`, `parentSuite`, `suite`, `subSuite`, `epic`, `feature`, `story`, `framework`, `language`, `layer`, `thread`, `host`, `severity`, `tag`, `owner`, `lead`.

Based on gut feeling, I think this is the case:

| Label        | Type |
|--------------|------|
| package      | one  |
| testClass    | one  |
| testMethod   | one  |
| parentSuite  | one  |
| suite        | one  |
| subSuite     | one  |
| epic         | many |
| feature      | many |
| story        | one? |
| framework    | N/A  |
| language     | N/A  |
| layer        | N/A  |
| thread       | one  |
| host         | N/A  |
| severity     | one  |
| tag          | many |
| owner        | one  |
| lead         | N/A  |

Could you please provide some comments regarding this? 🙏

**Comment by baev** [comment, marked as answer]
There are no restrictions on label duplicates in Allure. Although, for some of the labels, only the first one is considered. 

( Processing the first one and not the last one was a poor design decision since it limits the possibilities to override labels, see https://github.com/allure-framework/allure-php-commons2/pull/57, for example )

Besides that, some frameworks do additional logic on top of it, like in PyTest -- default suite labels (provided by PyTest reporter) will be only added to the results if no user-provided value is found. Although, users are free to add multiple suite labels if they wish. Such kind of behaviours are unclear and very hard to fix since the change will break a lot of tests for Allure Report users. 

As for now, the plan is to align the way we work with labels across frameworks and then address this particular issue in the next major Allure release. 

Besides that, Allure TestOps provides a way to change how labels are processed by creating rules (where the label value could be mapped to other entities such as issues, custom fields etc). That also works with system labels. 

In Allure 3, we also will introduce user-defined mappings (via configuration files) for labels (and not only the labels).

Meanwhile, regarding your question:

| Label | Type | Allure Report | Allure TestOps | Comment |
| ----- | ----- | ----- | ----- | ----- | 
| ALLURE_ID     | first  | ❌ | ✅ | 
| AS_ID     | first  | ❌ | ✅ | Deprecated, use `ALLURE_ID` instead |
| ALLURE_MANUAL     | first  | ❌ | ✅ | 
| package      | first   |  ✅  | ❌ | 
| testClass    | first  | ✅  | ✅ | 
| testMethod   | first  | ✅  | ✅ | 
| parentSuite  | all  | ✅  | ✅ | 
| suite        | all  | ✅  | ✅ | 
| subSuite     | all  | ✅  | ✅ | 
| epic         | all | ✅  | ✅ | 
| feature      | all | ✅  | ✅ | 
| story        | all | ✅  | ✅ | 
| framework    | first | ✅  | ❌ | No default processing for Allure TestOps, but often used in user-defined mappings |
| language     | first  | ✅  | ❌ | No default processing for Allure TestOps, but often used in user-defined mappings |
| layer        | first  | ✅  | ✅ | 
| thread       | first  | ✅  | ✅ | 
| host         | first  | ✅  | ✅ | 
| severity     | first  | ✅  | ❌ | 
| tag          | all | ✅  | ✅ | 
| owner        | all / first | ✅ first | ✅ all | 
| lead         | all  | ❌ | ✅ |

**Reply by noomorph** [reply]
Just a minor correction - owner is "first" in Allure reports.
